[Setting] MainTitleView 컴포넌트 추가

tea-hkim opened the pull request:
MainTitleView라는 컴포넌트를 추가하였습니다

## 작업내용
<!-- 작업 내용과 이미지를 첨부해주세요. -->
### MainTitleView를 컴포넌트로 만들었습니다
    - 내부자체적으로 layout이 설정되어있습니다
    - title로 쓰고 싶은 문자열을 입력해주면 됩니다
    - subtitle의 경우 없는 뷰도 있을 것 같아서 타입을 옵셔널로, 생성자에 기본값을 nil로 설정하였습니다
    - 퀵헬프를 추가하였습니다
 

|작업 후|
|:---:|
|<img width="250" src="https://user-images.githubusercontent.com/81206228/196868123-db1b3a0a-6e0e-4934-b62b-f7c6a9f06e29.png">|


## 리뷰포인트
<!-- 리뷰가 필요한 포인트와 해당 되는 커밋을 링크로 걸어주세요. -->

<img width="250" alt="image" src="https://user-images.githubusercontent.com/81206228/196868575-2b075bf0-ee5b-406e-b416-2e7d38c48632.png">

- title 자체에 상단 패딩이 들어가 있어서 MainTitleView의 constraint를 맞추실 때 view의 safe area에 맞춰주시면 됩니다!
- 좌우 constraints는 view에 맞춰주시면 됩니다

## Checklist
- [x] 브랜치를 가져와 작업한 경우 이전 브랜치에 PR을 보냈는지 확인
- [x] 빌드를 위해 SceneDelegate 수정한 것 PR로 올리지 않았는지 확인
- [x] 필요없는 주석, 프린트문 제거했는지 확인
- [x] 컨벤션 지켰는지 확인
- [x] final, private 제대로 넣었는지 확인
- [x] 다양한 디바이스에 레이아웃이 대응되는지 확인
  - [x] iPhone SE
  - [x] iPhone 13
  - [x] iPhone 13 Pro Max

tea-hkim committed:
Setting:gear:: MainTitleView 컴포넌트 추가

MainTitleView라는 컴포넌트를 추가하였습니다
  - 내부자체적으로 layout이 설정되어있습니다
  - title로 쓰고 싶은 문자열을 입력해주면 됩니다
  - subtitle의 경우 없는 뷰도 있을 것 같아서 타입을 옵셔널로, 생성자에 기본값을 nil로 설정하였습니다

tea-hkim committed:
Chore:broom:: mainTitle 퀵헬프 추가

Lia316 commented on Projects/DesignSystem/Sources/Component/MainTitleView.swift:
```diff
@@ -0,0 +1,60 @@
+    private func createLayout() {
+        addSubviews([titleLabel, subtitleLabel])
+        
+        titleLabel.snp.makeConstraints { make in
+            make.top.equalToSuperview().inset(40)
+            make.horizontalEdges.equalToSuperview().inset(20)
```
[P1]

title의 top inset은 30으로 알고 있습니다?
<img width="253" alt="image" src="https://user-images.githubusercontent.com/73650994/196925576-65d33c53-e9ed-46f9-a52d-4a7713a38c6e.png">

Chaeho-Min replied:
닉이 억울할거 같은 포인트...
닉이 PR 올린 후에 바뀌었답니다😅

Lia316 replied:
저런~! 일하세요 닉🦊

chaneeii replied:
poor nick..

<img width="300" src="https://user-images.githubusercontent.com/63157395/196941304-4d786562-be2b-43cf-b2ad-c9fbe9de2880.png">

tea-hkim replied:
ec55e9b 변경했습니다!

Lia316 commented on Projects/DesignSystem/Sources/Component/MainTitleView.swift:
```diff
@@ -0,0 +1,60 @@
+            make.horizontalEdges.equalToSuperview().inset(20)
+        }
+        
+        subtitleLabel.snp.makeConstraints { make in
+            make.top.equalTo(titleLabel.snp.bottom).offset(12)
+            make.horizontalEdges.equalToSuperview().inset(20)
+        }
```
[P2]

auto layout은 x, y, width, height를 모두 추론할 수 있게 해줘야 제대로 레이아웃을 잡아줍니다.
현재 bottom을 명시하지 않아서 height를 추론할 수 없는 상태에요!

ScrollView 같이 높이가 중요한 뷰에서는 이것 때문에 레이아웃이 원하는 대로 안 잡힐 수 있으니 수정 요청드립니다!

chaneeii replied:
p2)
동의합니다.

근데 타이틀 아래 bottom 이 어떻게 될지 모르는데, 어떻게 컴포넌트를 작성해야할까요..? @Lia316 
제가 생각나는 방법은 최대 높이 2줄의 높이로 일단 잡아두는것...??

@Chaeho-Min 
- subtitle 내용이 2줄이 넘지 않아야할것같네요 / 2줄일때 높이를 알수있을까요?

Lia316 replied:
@chaneeii 
어차피 MainTitleView도 상위 뷰에서 레이아웃을 잡아줘야하니 superview와 동일한 bottom을 잡아도 될 것 같아요!
그리고 greaterThanOrEqualTo를 사용해서 최소 높이를 정해주고,
line 수를 적절하게 설정해주면 되지 않을까 싶습니당

Chaeho-Min replied:
오토레이아웃... 어렵네요
타이틀 한 줄 & 설명 한 줄일 때를 최소 높이로 해서 이 값을 드리면 되는건가요?

Chaeho-Min replied:
오토레이아웃... 어렵네요
타이틀 한 줄 & 설명 한 줄일 때를 최소 높이로 해서 이 값을 드리면 되는건가요?

tea-hkim replied:
아하 제가 의도한 건 이 컴포넌트를 사용하는 사람이 MainTitleView의 높이와 너비를 직접 지정할 수 있도록 하는 것입니다!

유동적으로 높이가 변할 수 있도록 bottom도 지정해 두지 않았고, 
Hierarchy에서 constraints 관련 에러가 발생하지 않아 PR을 보냈는데 
지금 생각해 보니까 문제가 있네요...

286e2d7 변경하였습니다!

1. greaterThanOrEqualTo를 아보가 피그마 컴포넌트에 만들어 둔 94 포인트로 변경했습니다
2. MainTitleView를 base로 stackView를 올려 높이가 유동적으로 변하도록 하였습니다
3. stackView를 사용한 이유는 greaterThanOrEqualTo으로 했을 경우 높이를 명확하게 인식하지 못하기 때문입니다!

Lia316 commented on Projects/DesignSystem/Sources/Component/MainTitleView.swift:
```diff
@@ -0,0 +1,60 @@
+    private let titleLabel = UILabel()
+    private let subtitleLabel = UILabel()
+    
+    public init(title: String, subtitle: String? = nil) {
+        super.init(frame: .zero)
+        titleLabel.text = title
+        subtitleLabel.text = subtitle
```
[Q]

subtitle을 옵셔널로 받는데, 
subtitle이 없을 경우에도 이 친구의 레이아웃을 잡으신 건 의도하신 걸까요?
그러면 title만 있는데 아래 패딩이 너무 많이 잡히지 않나 싶어서 여쭤봅니다~!

Chaeho-Min replied:
오토레이아웃... 어렵네요
타이틀 한 줄 & 설명 한 줄일 때를 최소 높이로 해서 이 값을 드리면 되는건가요?

tea-hkim replied:
아하 제가 의도한 것은 subTitle에 아무런 값을 넘겨주지 않을 경우,
높이가 없는 UILable의 text는 빈 문자열을 할당 받게 되어 화면에 보이지 않고 고유의 공간을 가지지 않는다는 것인데,

그렇게 되면 화면에는 보이지 않지만 메모리에는 올라가 있는 상태가 되겠네요 ㅠ

수정하겠습니다!

tea-hkim replied:
286e2d7 변경하였습니다!

Lia316 reviewed (changes requested):
모두가 쓰는 컴포넌트를 생성해주셔서 감사드려요!! 😆

모두가 사용하는 만큼 레이아웃에 대해 깐깐하게 봤어요 ㅎ__ㅎ

Chaeho-Min commented on Projects/DesignSystem/Sources/Component/MainTitleView.swift:
```diff
@@ -0,0 +1,60 @@
+        titleLabel.numberOfLines = 0
+        titleLabel.font = .systemFont(ofSize: 26, weight: .heavy)
+        
+        subtitleLabel.textColor = .zestyColor(.gray8A8A8E)
```
p2;
<img width="740" alt="image" src="https://user-images.githubusercontent.com/75792767/196955392-90ff21f3-8bac-45ac-8445-c0d3283aa9b4.png">
텍스트 컬러는 UIColor에서 기본으로 제공해주는 애들을 사용할 것 같아서 이 텍스트 컬러는 따로 추가할 필요 없이 `.secondaryLabel`을 사용해주세요!

tea-hkim replied:
167d0e8 변경했습니다!

Chaeho-Min commented on Projects/DesignSystem/Sources/Component/MainTitleView.swift:
```diff
@@ -0,0 +1,60 @@
+        
+        subtitleLabel.textColor = .zestyColor(.gray8A8A8E)
+        subtitleLabel.numberOfLines = 0
+        subtitleLabel.font = .systemFont(ofSize: 16, weight: .semibold)
```
p1;
<img width="622" alt="image" src="https://user-images.githubusercontent.com/75792767/196956036-0a5cf823-cb4a-46cb-a0ec-51e047e44fe6.png">
subtitle 폰트의 weight은 .regular입니다! semibold로 설정하신 이유가 있나요?

tea-hkim replied:
피그마에 랜더링되는 서체의 굵기가 시뮬레이터나 디바이스에 랜더링 될 때와 차이가 있어서 직접 맞췄습니다!
디자인대로 변경했습니다! ea24c60

Chaeho-Min replied:
오 얼마나 차이나는지 한 번 보고 싶네요 👀

Chaeho-Min reviewed (changes requested):
컴포넌트 추가 감사드립니다🙇‍♂️
그런데 왠지 간격에 수정이 있을 것 같은 불길한 예감😅
남겨놓은 코멘트만 확인해주세요!

tea-hkim committed:
Chore:broom:: Color Asset gray8A8A8E 삭제

gray8A8A8E를 secondaryLabel로 변경했습니다

tea-hkim committed:
Chore:broom:: font weight 변경

디자인대로 수정했습니다

tea-hkim committed:
Design:art:: title inset 40에서 30으로 변경

Chaeho-Min reviewed (approved):
제 코멘트는 다 해결돼서 어프루브요~

tea-hkim committed:
Chore:broom:: subtitle 예외 처리

subtitle을 예외 처리 하였습니다
- subtitle에 값이 들어오지 않을 경우 빈 문자열이 리턴됩니다.
- 빈 문자열이 들어올 경우 subtitle의 layout이 잡히지 않도록 했습니다

Lia316 commented on Projects/DesignSystem/Sources/Component/MainTitleView.swift:
```diff
@@ -0,0 +1,88 @@
+            make.height.greaterThanOrEqualTo(31)
+        }
+        
+        if !subtitle.isEmpty {
+            titleStackView.addArrangedSubview(subtitleLabel)
```
[P1]

마찬가지로 subtitle이 nil일 경우에 대한 처리가 필요합니다.
if let 등의 바인딩 부탁드려요!

Lia316 commented on Projects/DesignSystem/Sources/Component/MainTitleView.swift:
```diff
@@ -0,0 +1,88 @@
+        titleLabel.text = title
+        subtitleLabel.text = subtitle
+        
+        guard let subtitle = subtitle else {return}
+        
+        configureUI(subTitle: subtitle)
+        createLayout(subtitle: subtitle)
```
[P1]

현재 작성하신 코드는 subtitle이 없으면 아예 레이아웃이 그려지지 않습니다.

tea-hkim replied:
기본값을 주고도 옵셔널 타입으로 잘못 설정을 했습니다 ㅠ 
181baf7 변경하였습니다!

Lia316 commented on Projects/DesignSystem/Sources/Component/MainTitleView.swift:
```diff
@@ -0,0 +1,88 @@
+/// 높이는 최소 94 point로 정해져 있습니다.
+///
+/// top constraint와 너비만 지정해주면 됩니다
+public final class MainTitleView: UIView {
+    private let titleLabel = UILabel()
+    private let subtitleLabel = UILabel()
+    private let titleStackView = UIStackView()
+    
+    public init(title: String, subtitle: String? = "") {
```
[P3]

UILabel처럼 나중에도 타이틀과 서브 타이틀을 수정할 수 있게 해주면 좋을 것 같습니다!
컴포넌트의 경우에는 디자인에 따라 유동성이 있어도 좋다고 생각해요

```suggestion
public final class MainTitleView: UIView {
    public var titleLabel = UILabel()
    public var subtitleLabel = UILabel()
    private let titleStackView = UIStackView()
    
    public init(title: String = "", subtitle: String? = "") {
```

tea-hkim replied:
4eba8f5 변경하였습니다!

생성자의 경우 이 컴포넌트의 역할이 title을 만든는 것이기 때문에 title에는 기본값을 두지 않았습니다.
이 컴포넌트를 사용하는 개발자에게 컴포넌트의 역할과 컴포넌트의 동작인 title 입력을 해야 한다는 것을 명시하기 위함입니다!

Lia316 replied:
여전히 자유를 더 줘도 괜찮다고는 생각하지만,,
기본 컴포넌트와 다르게 레이아웃이 잡혀있는 상태이니 title 입력을 강제하는 게 낫다는 생각도 드네요
반영 감사합니다!

Lia316 reviewed (changes requested):
수고하셨습니다!

Lia316 reviewed (commented):
리뷰 수정했습니다!

tea-hkim committed:
Chore:broom:: subtitle을 옵셔널에서 문자열 타입으로 변경

옵셔널 타입의 경우 nil이 들어와서 꺼지는 autoLayout을 못잡는 경우가 있어
문자열 타입으로 변경했습니다!

tea-hkim committed:
Chore:broom:: title, subtitle 접근권한 변경

디자인의 유동성을 위해 접근 권한을 변경합니다

tea-hkim committed:
Merge branch 'dev' into feat/Component/mainTitle
